File: Leetcoding-Actions/Explore-Learn/Arrays/DuplicateZeros.java
class Solution {
    public void duplicateZeros(int[] arr) {
        /**
        * T: O(n) and S: O(n)
        */
        int i;
        for (i = 0; i < arr.length; i++) {
            if (arr[i] == 0) break;
        }
        Set<Integer> zerosPosition = new HashSet<>();
        ArrayDeque<Integer> nonZeroQ = new ArrayDeque<>();
        int zerosCount = 0;
        for (int j = i; j < arr.length; j++) {
            if (arr[j] == 0) {
                zerosPosition.add(j + zerosCount);
                zerosPosition.add(j + zerosCount + 1);
                zerosCount++;
            } else {
                nonZeroQ.add(arr[j]);
            }
        }
        for (int j = i; j < arr.length; j++) {
            if (zerosPosition.contains(j)) {
                arr[j] = 0;
            } else {
                arr[j] = nonZeroQ.removeFirst();
            }
        }
    }
}

File: Leetcoding-Actions/Explore-Learn/Arrays/DuplicateZeros.py
class Solution:
    def duplicateZeros(self, arr: List[int]) -> None:
        """
        Do not return anything, modify arr in-place instead.
        T: O(n) and S: O(n)
        """
        # Position of first 0
        for i in range(len(arr)):
            if arr[i] == 0: break
        
        # Zero positions and nonZero elements
        zeroPosition = set()
        nonZeroQ = deque()
        zerosCount = 0
        for j in range(i, len(arr)):
            if arr[j] == 0:
                zerosCount += 1
                zeroPosition.add(j + zerosCount - 1)
                zeroPosition.add(j + zerosCount - 1 +1)
            else:
                nonZeroQ.append(arr[j])
        
        # Put 0 at zeroPosition and put nonZero elements at other position 
        for j in range(i, len(arr)):
            if j in zeroPosition:
                arr[j] = 0
            else:
                arr[j] = nonZeroQ.popleft()
